walk/walk.go

```diff
@@ -399,22 +402,37 @@ func buildTrie(c *config.Config, updateRels *UpdateFilter, isIgnored isIgnoredFu
 	// An error group to handle error propagation
 	eg := errgroup.Group{}
 	eg.Go(func() error {
-		return walkDir(c.RepoRoot, "", &eg, limitCh, updateRels, isIgnored, trie)
+		return walkDir(c, c.ValidBuildFileNames, "", &eg, limitCh, updateRels, isIgnored, trie)
 	})
 
 	return trie, eg.Wait()
 }
 
 // walkDir recursively and concurrently descends into the 'rel' directory and builds a trie
-func walkDir(root, rel string, eg *errgroup.Group, limitCh chan struct{}, updateRels *UpdateFilter, isIgnored isIgnoredFunc, trie *pathTrie) error {
+func walkDir(c *config.Config, validBuildFileNames []string, rel string, eg *errgroup.Group, limitCh chan struct{}, updateRels *UpdateFilter, isIgnored isIgnoredFunc, trie *pathTrie) error {
 	limitCh <- struct{}{}
 	defer (func() { <-limitCh })()
 
-	entries, err := os.ReadDir(filepath.Join(root, rel))
+	// Absolute path to the directory being visited
+	dir := filepath.Join(c.RepoRoot, rel)
+
+	entries, err := os.ReadDir(dir)
 	if err != nil {
 		return err
 	}
 
+	trie.build, trie.buildFileErr = loadBuildFile(c, validBuildFileNames, rel, dir, entries)
+
+	if trie.build != nil {
+		// Update the `validBuildFileNames` config while traversing the tree.
+		for _, d := range trie.build.Directives {
+			if d.Key == "build_file_name" {
```

Comment:
This is an example of directives that effect the fs-walk. Those directives must be read while we actually do the concurrent fs-walk and later during `Configure`.

IMO this shows how some builtin "core options" are different then "language directives". The [poc moving exclude to fs-walk](https://github.com/bazel-contrib/bazel-gazelle/pull/1928/files#r1962409426) highlights this distinction even more.

Reply:
This duplication of logic is unfortunate. It makes me wonder how much performance parsing BUILD files in parallel gives us and whether it's worth it. It should be an I/O and memory bound process, not CPU, but there's still probably some speedup there.

Instead of passing in `validBuildFileNames`, how about passing down `*walk.Configurer` as an argument and applying that to the configuration in each directory? Then we can use the real code. We should be able to get `*walk.Configurer` out of `c.Exts`. This requires more allocation (for cloning the `*config.Config` in each directory). Without a benchmark, it's hard to say what the performance impact is though.

Reply:
I have not properly measured the performance yet. Parsing concurrently is probably negligible, the I/O concurrently might make a minor difference, the only significant thing might be when `# gazelle:exclude`s exclude a significant amount of directories which will reduce CPU + I/O.

Performance aside... I think your idea [changing the trie to a concurrent cache](https://github.com/bazel-contrib/bazel-gazelle/issues/1891#issuecomment-2666792101) would require something similar to this PR? I was imagining that cache would essentially make all this `buildTrie` + `loadBuildFile` lazy+cached, but those two steps need to be merged together before they can be loaded+cached together...

Reply:
Since we're not yet measuring performance, and the allocation cost is likely to be small, how about pulling in `*walkConfig` here? I see you have it in the `# gazelle:exclude` PoC. Let's not do all of that yet in this PR, but let's move in that direction instead of duplicating the configuration code.

Reply:
I agree `walkConfig` should move here. It will just require a bit of refactoring that I was hoping to avoid for now, I might try doing that in a separate PR that can land first... 🤔

Reply:
I think this is a good prefactor? https://github.com/bazel-contrib/bazel-gazelle/pull/2044

Reply:
> how about pulling in *walkConfig here

I've added a commit doing that, and rebased [the branch moving exclude](https://github.com/jbedard/bazel-gazelle/compare/loadbuild-concurrent...jbedard:bazel-gazelle:fs-exclude)